File: depth_fisrt_search/depth_first_search.py
# Complexity O(v + e) - v - number of vertices, e - number of edges (need to write name to array + iterate thru the children)
# Space O(v) - length of returning array
class Node:
    def __init__(self, name):
        self.children = []
        self.name = name

    def addChild(self, name):
        self.children.append(Node(name))
        return self

    def depthFirstSearch(self, array):
        array.append(self.name)
        for child in self.children:
            child.depthFirstSearch(array)
        return array


if __name__ == '__main__':
    graph = Node("A")
    graph.addChild("B").addChild("C").addChild("D")
    graph.children[0].addChild("E").addChild("F")
    graph.children[2].addChild("G").addChild("H")
    graph.children[0].children[1].addChild("I").addChild("J")
    graph.children[2].children[0].addChild("K")
    print(graph.depthFirstSearch([]), "vs", ["A", "B", "E", "F", "I", "J", "C", "D", "G", "K", "H"])

File: depth_fisrt_search/depthFirstSearch.go
package main

import "fmt"

type Node struct {
	Children	[]Node
	Name		string
}

func (n *Node) addChild(name string) *Node {
	n.Children = append(n.Children, Node{Children: nil, Name: name})
	return n
}

func (n Node) depthFirstSearch(array *[]string) []string {
	*array = append(*array, n.Name)
	for _, child := range n.Children {
		child.depthFirstSearch(array)
	}
	return *array
}

func main() {
	graph := Node{Children: nil, Name: "A"}
	graph.addChild("B").addChild("C").addChild("D")
	graph.Children[0].addChild("E").addChild("F")
	graph.Children[2].addChild("G").addChild("H")
	graph.Children[0].Children[1].addChild("I").addChild("J")
	graph.Children[2].Children[0].addChild("K")

	var array []string
	var ex = []string{"A", "B", "E", "F", "I", "J", "C", "D", "G", "K", "H"}
	fmt.Println(graph.depthFirstSearch(&array), "vs", ex)
}
